### PR TITLE
am: rewrite for multiprocess support

### PR DESCRIPTION
**Pull request by liamwhite**
This is all of the core and am changes--except for surface capture and use of the guest applets--extracted from #12756. This is testable standalone and should ideally not change any application behavior.

### CODE REVIEW

File: src/core/hle/service/ns/ns.cpp [hunk 1]
```diff
@@ -554,6 +555,43 @@ IDocumentInterface::IDocumentInterface(Core::System& system_)
 
 IDocumentInterface::~IDocumentInterface() = default;
 
+void IDocumentInterface::GetApplicationContentPath(HLERequestContext& ctx) {
+    LOG_ERROR(Service_NS, "(STUBBED) called");
+    // inputs ContentPath
+    // returns boolean and fs::Path as fixed size out buffer
```
Comment (FearlessTobi) [inline src/core/hle/service/ns/ns.cpp:561]: this does not seem to match what you are actually returning

File: src/core/hle/service/ns/ns.cpp [hunk 2]
```diff
@@ -554,6 +555,43 @@ IDocumentInterface::IDocumentInterface(Core::System& system_)
 
 IDocumentInterface::~IDocumentInterface() = default;
 
+void IDocumentInterface::GetApplicationContentPath(HLERequestContext& ctx) {
+    LOG_ERROR(Service_NS, "(STUBBED) called");
```
Comment (FearlessTobi) [inline src/core/hle/service/ns/ns.cpp:559]: LOG_WARNING

File: src/core/hle/service/ns/ns.cpp [hunk 3]
```diff
@@ -554,6 +555,43 @@ IDocumentInterface::IDocumentInterface(Core::System& system_)
...
+    // inputs ContentPath
+    // returns boolean and fs::Path as fixed size out buffer
+
+    IPC::ResponseBuilder rb{ctx, 2};
+    rb.Push(ResultUnknown);
+}
+
+void IDocumentInterface::ResolveApplicationContentPath(HLERequestContext& ctx) {
+    LOG_ERROR(Service_NS, "(STUBBED) called");
```
Comment (FearlessTobi) [inline src/core/hle/service/ns/ns.cpp:568]: remove this

File: src/core/hle/service/ns/ns.cpp [hunk 4]
```diff
@@ -554,6 +555,43 @@ IDocumentInterface::IDocumentInterface(Core::System& system_)
...
+
+    struct ContentPath {
+        u8 file_system_proxy_type;
+        u64 program_id;
+    };
+    static_assert(sizeof(ContentPath) == 0x10, "ContentPath has wrong size");
+
+    IPC::RequestParser rp{ctx};
+    auto content_path = rp.PopRaw<ContentPath>();
+    LOG_ERROR(Service_NS, "(STUBBED) called, file_system_proxy_type={}, program_id={:016X}",
```
Comment (FearlessTobi) [inline src/core/hle/service/ns/ns.cpp:578]: LOG_WARNING

File: src/core/hle/service/ns/ns.cpp [hunk 5]
```diff
@@ -554,6 +555,43 @@ IDocumentInterface::IDocumentInterface(Core::System& system_)
...
+              content_path.file_system_proxy_type, content_path.program_id);
+
+    IPC::ResponseBuilder rb{ctx, 2};
+    rb.Push(ResultSuccess);
+}
+
+void IDocumentInterface::GetRunningApplicationProgramId(HLERequestContext& ctx) {
+    IPC::RequestParser rp{ctx};
+    const auto caller_program_id = rp.PopRaw<u64>();
+    LOG_INFO(Service_NS, "(STUBBED) called, caller_program_id={:016X}", caller_program_id);
```
Comment (FearlessTobi) [inline src/core/hle/service/ns/ns.cpp:588]: LOG_WARNING

**Review by FearlessTobi — COMMENTED** [review]
I took a quick glance and did not see any major issues.
Nice work :)

File: src/core/hle/service/ns/ns.cpp (hunk 1 above)
Comment (liamwhite) [reply src/core/hle/service/ns/ns.cpp:561]: I wanted to document the arguments for this method but I did not need to actually implement it. What should I do in this case?

**Review by german77 — APPROVED** [review]

**Pull request merged**